beacon-chain/operations/slashings/service.go

```diff
@@ -0,0 +1,207 @@
+package slashings
+
+import (
+	"context"
+	"sort"
+	"sync"
+
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	beaconstate "github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	"github.com/prysmaticlabs/prysm/shared/sliceutil"
+)
+
+// Pool implements a struct to maintain pending and recently included voluntary exits. This pool
+// is used by proposers to insert into new blocks.
+type Pool struct {
+	lock                    sync.RWMutex
+	pendingProposerSlashing []*ethpb.ProposerSlashing
+	pendingAttesterSlashing []*PendingAttesterSlashing
+	included                map[uint64]bool
+}
+
+// PendingAttesterSlashing represents an attester slashing in the operation pool.
+// Allows for easy binary searching of included validator indexes.
+type PendingAttesterSlashing struct {
+	attesterSlashing *ethpb.AttesterSlashing
+	validatorToSlash uint64
+}
+
+// NewPool accepts a head fetcher (for reading the validator set) and returns an initialized
+// slashed validator pool.
+func NewPool() *Pool {
+	return &Pool{
+		pendingProposerSlashing: make([]*ethpb.ProposerSlashing, 0),
+		pendingAttesterSlashing: make([]*PendingAttesterSlashing, 0),
+		included:                make(map[uint64]bool),
+	}
+}
+
+// PendingAttesterSlashings returns exits that are ready for inclusion at the given slot. This method will not
+// return more than the block enforced MaxAttesterSlashings.
+func (p *Pool) PendingAttesterSlashings() []*ethpb.AttesterSlashing {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+
+	var included []uint64
+	pending := make([]*ethpb.AttesterSlashing, 0)
+	for i, slashing := range p.pendingAttesterSlashing {
+		if i >= int(params.BeaconConfig().MaxAttesterSlashings) {
+			break
+		}
+		if sliceutil.IsInUint64(slashing.validatorToSlash, included) {
+			continue
+		}
+		attSlashing := slashing.attesterSlashing
+		slashedVal := sliceutil.IntersectionUint64(attSlashing.Attestation_1.AttestingIndices, attSlashing.Attestation_2.AttestingIndices)
+		// Doesn't need to be sorted.
+		included = append(included, slashedVal...)
+		pending = append(pending, attSlashing)
+	}
+
+	return pending
+}
+
+// PendingProposerSlashings returns proposer slashings that are ready for inclusion at the given slot.
+// This method will not return more than the block enforced MaxProposerSlashings.
+func (p *Pool) PendingProposerSlashings() []*ethpb.ProposerSlashing {
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	pending := make([]*ethpb.ProposerSlashing, 0)
+	for i, slashing := range p.pendingProposerSlashing {
+		if i >= int(params.BeaconConfig().MaxProposerSlashings) {
+			break
+		}
+		pending = append(pending, slashing)
+	}
+	return pending
+}
+
+// InsertAttesterSlashing into the pool. This method is a no-op if the pending exit already exists,
```

Comment:
not `pending exit`, please edit that

Reply:
Apologies, it was initially copy pasted from exits.